### PR TITLE
Save current mask when leaving the scene

### PR DESCRIPTION
**Pull request by louist103**
Saves the current mask when leaving and loading a new scene and on resets.
Had to do some sketchy things but hey it works.

### CODE REVIEW

**Comment by ETR-BTF** [comment]
Does this also keep your mask equipped when entering bodies of water or when climbing ladders? Iirc those actions also unequip masks.

**Comment by louist103** [comment]
> Does this also keep your mask equipped when entering bodies of water or when climbing ladders? Iirc those actions also unequip masks.

They shouldn't be effected unless somehow spawning in either of those two states messes with this. I can test it later today.

**Comment by louist103** [comment]
> Does this also keep your mask equipped when entering bodies of water or when climbing ladders? Iirc those actions also unequip masks.

Entering water doesn't remove it but you can't equip and remove them. This appears to be the same as vanilla.

File: ASM/src/player_save_mask.asm [hunk 1]
```diff
@@ -0,0 +1,6 @@
+player_save_mask:
+    lbu     t0, 0x3B(s2) ;saveContext->savedMask (custom field)
+    sb      t0, 0x014F(a1) ;player->currentMask
+
+    j       0x8038C9A0 ; Pre relocated jump. Player is always in the same place so it *should* be fine.
```
Comment (mracsys) [inline ASM/src/player_save_mask.asm]: I would avoid hard-coding addresses if possible. See #1622

Comment (louist103) [reply ASM/src/player_save_mask.asm]: I don't like it either but between not having a linker map or similar and having to deal with relocations I don't know of a better way right now.

Comment (louist103) [reply ASM/src/player_save_mask.asm]: Alright I found a way to do this properly. It now no longer uses a hard coded address and properly relocates the function.

**Comment by cjohnson57** [comment]
I'm a little unsure about this one with the "sketchy stuff" done... Especially since, now that we have masks on D pad, re-equipping isn't much of an issue

**Comment by louist103** [comment]
> I'm a little unsure about this one with the "sketchy stuff" done... Especially since, now that we have masks on D pad, re-equipping isn't much of an issue

Well everything done has a good reason and needs to be done. The zeroing of the reloc to prevent an address of a function that is in a constant place from being relocated, relocating one of the overlay functions my self (though to be super correct we should make a `relocate_ovl_func` function). I will agree that writing to an empty slot in save context is sketchy though.

**Comment by cjohnson57** [comment]
Hm, alright. Doing a once over of the code and notes, there doesn't appear to be any other hacks using this slot. If you can real quick note the slot and what you're doing with it in the document Notes/extra-save-data.md, it should be good to go. Thanks!

**Comment by ETR-BTF** [comment]
Is there a fail-safe for when a mask cannot be unequipped when it should be unequipped? For example, you can equip swap a mask as adult and then lose the ability to equip swap masks. I believe that this PR also allows masks in places where normally you can't equip masks, such as in boss rooms. Is this desirable?

**Pull request merged**